### PR TITLE
Backend tests

### PR DESCRIPTION
**Pull request by Crash1212**

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -1,4 +1,4 @@
-MONGO_URL=
+MONGO_URL=mongodb+srv://bandvov:z8tVhud7CW7v3rRK@cluster0-f8dbj.azure.mongodb.net/horondi-test?retryWrites=true&w=majoritys
```
Comment (VolodymyrTrach) [inline .env.example:1]: You should not leave it here. Make an example of it

File: src/config/db.js [hunk 2]
```diff
@@ -4,6 +4,7 @@ const configService = require('../utils/configService');
 
 const connectDB = async () => {
   const db = await configService.getSecret('MONGO_URL');
+  // const db = process.env.TEST_MONGO
```
Comment (VolodymyrTrach) [inline src/config/db.js:7]: don't leave commented code

File: src/config/db.js [hunk 3]
```diff
@@ -13,6 +14,7 @@ const connectDB = async () => {
     });
     console.log('MongoDB Connected...');
   } catch (err) {
+    console.log('Mongo connection error')
```
Comment (VolodymyrTrach) [inline src/config/db.js:17]: remove console.log

File: src/modules/currency/currency.graphql.js [hunk 4]
```diff
@@ -1,13 +1,13 @@
 const currencyType = `type Currency {
     _id: ID!
-    date: String!
+    date: String
```
Comment (VolodymyrTrach) [inline src/modules/currency/currency.graphql.js:3]: why it is not required anymore?

File: src/modules/currency/currency.graphql.js [hunk 5]
```diff
@@ -1,13 +1,13 @@
...
     convertOptions: [ConvertOption!]
 }
 `;
 
 const currencyInput = `
     input CurrencyInput {
-    date: String!
+    date: String
```
Comment (VolodymyrTrach) [inline src/modules/currency/currency.graphql.js:10]: why it is not required anymore?

Comment (Crash1212) [reply src/modules/currency/currency.graphql.js:10]: we do not have to enter the date manually

File: src/config/db.js (hunk 3 above)
Comment (Crash1212) [reply src/config/db.js:17]: this is dev console log to get info about db status

File: src/modules/currency/currency.graphql.js (hunk 4 above)
Comment (Crash1212) [reply src/modules/currency/currency.graphql.js]: fixed

File: src/config/db.js (hunk 2 above)
Comment (Crash1212) [reply src/config/db.js]: fixed

File: .env.example (hunk 1 above)
Comment (Crash1212) [reply .env.example]: fixed

File: src/config/db.js (hunk 3 above)
Comment (VolodymyrTrach) [reply src/config/db.js:16]: ок

File: package.json [hunk 6]
```diff
@@ -21,24 +21,29 @@
...
     "@azure/keyvault-secrets": "^4.0.4",
     "@hapi/joi": "^17.1.1",
+    "apollo-boost": "^0.4.9",
+    "apollo-cache-inmemory": "^1.6.6",
+    "apollo-link-http": "^1.5.17",
     "apollo-server": "^2.14.5",
     "azure": "^2.3.1-preview",
     "bcryptjs": "^2.4.3",
     "config": "^3.2.5",
-    "cors": "^2.8.5",
```
Comment (Alan18081) [inline package.json:28]: Why did you remove the `cors` package?

File: package.json [hunk 7]
```diff
@@ -21,24 +21,29 @@
...
+    "cross-fetch": "^3.0.5",
     "dotenv": "^8.2.0",
-    "graphql": "^15.1.0",
+    "graphql": "^15.2.0",
     "jsonwebtoken": "^8.5.1",
     "mongoose": "^5.9.18",
     "mongoose-middleware": "^2.0.1",
     "morgan": "^1.9.1",
+    "node-fetch": "^2.6.0",
```
Comment (Alan18081) [inline package.json:38]: Why do we need `node-fetch` if we already have axios?

File: package.json [hunk 8]
```diff
@@ -21,24 +21,29 @@
...
     "@hapi/joi": "^17.1.1",
+    "apollo-boost": "^0.4.9",
+    "apollo-cache-inmemory": "^1.6.6",
+    "apollo-link-http": "^1.5.17",
     "apollo-server": "^2.14.5",
     "azure": "^2.3.1-preview",
     "bcryptjs": "^2.4.3",
     "config": "^3.2.5",
-    "cors": "^2.8.5",
+    "cross-fetch": "^3.0.5",
```
Comment (Alan18081) [inline package.json:31]: Why do we need `node-fetch` if we already have `axios`?

File: src/__tests__/category/__snapshots__/category.query.test.js.snap [hunk 9]
```diff
@@ -0,0 +1,24 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
Comment (Alan18081) [inline src/__tests__/category/__snapshots__/category.query.test.js.snap:1]: Tests should not be in `src` directory, it should be on the same level as `src` folder. That's why we have `src` folder at all

File: src/__tests__/category/category.mutation.test.js [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+                        }
+                        available
+                    }
+                }
+            `
+        })
+
+        categoryId = res.data.addCategory._id
+        
+        expect(categoryService.addCategory({
```
Comment (Alan18081) [inline src/__tests__/category/category.mutation.test.js:64]: Please, avoid using direct service calls as much as possible. The idea of this kind of test is that you make HTTP call and check the response that is returned from the backend. But we should not care about inner workings of the services and their methods. Tests should know nothing about it. Of course, there are some exceptions, for example, when we need to delete entity to clear data that is used in tests but operation DELETE is not provided by API.

File: src/__tests__/category/category.query.test.js [hunk 11]
```diff
@@ -0,0 +1,62 @@
+/* eslint-disable no-undef */
+const { gql } = require('apollo-boost');
+const categoryService = require('../../modules/category/category.service')
+const categoryResolver = require('../../modules/category/category.resolver')
+const client = require('../../utils/apolloClient');
+
+require('dotenv').config();
+
+
+describe('querries', () => {
```
Comment (Alan18081) [inline src/__tests__/category/category.query.test.js:10]: Typo

File: src/__tests__/category/category.query.test.js [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    test('should recive all categories like in service and resolver', async () => {
```
Comment (Alan18081) [inline src/__tests__/category/category.query.test.js:11]: Typo

File: src/__tests__/category/category.mutation.test.js [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+const categoryResolver = require('../../modules/category/category.resolver')
+const client = require('../../utils/apolloClient');
+
+require('dotenv').config();
+
+let categoryId
+
+describe('mutations', () => {
+    test('should create new category', async () => {
+        const res = await client.mutate({
```
Comment (Alan18081) [inline src/__tests__/category/category.mutation.test.js:13]: When you get a response, please, check specific things
- data and all of its fields
- an array of errors (is it empty or does it contain any messages if you have provided invalid requests)
- HTTP status code if it is not already done by `apollo-client`

File: src/__tests__/category/category.mutation.test.js [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+            },
+            {
+                lang: "ukr",
+                value: "Тест"
+            }
+            ]
+        }})).resolves.toBe(res);
+    })
+
+    test('should delete category', async () => {
```
Comment (Alan18081) [inline src/__tests__/category/category.mutation.test.js:182]: Please, have also cases where you provide invalid data to check how business validation works, for example, if you try to delete the category with id that doesn't exist inside database

File: src/config/db.js [hunk 15]
```diff
@@ -13,6 +13,7 @@ const connectDB = async () => {
...
```
Comment (Alan18081) [inline src/config/db.js:16]: Please, add `err` as a second argument to the `console.log` to see the full error object

File: src/modules/category/category.service.js [hunk 16]
```diff
@@ -0,0 +1,30 @@
+const Category = require('./category.model');
```
Comment (Alan18081) [inline src/modules/category/category.service.js:1]: Please, move all newly added logic for categories to another merge request to merge it in development as quickly as possible.

File: src/modules/currency/currency.graphql.js (hunk 5 above)
Comment (Alan18081) [reply src/modules/currency/currency.graphql.js:10]: What is this date about? What real-world thing does it reflect?

File: src/modules/currency/currency.service.js [hunk 17]
```diff
@@ -0,0 +1,30 @@
+const Currency = require('./currency.model');
```
Comment (Alan18081) [inline src/modules/currency/currency.service.js:1]: Please, move all newly added logic for currencies to another merge request to merge it in development as quickly as possible.

File: src/modules/material/material.service.js [hunk 18]
```diff
@@ -0,0 +1,31 @@
+const Material = require('./material.model');
```
Comment (Alan18081) [inline src/modules/material/material.service.js:1]: Please, move all newly added logic for materials to another merge request to merge it in development as quickly as possible.

File: src/modules/pattern/pattern.service.js [hunk 19]
```diff
@@ -0,0 +1,30 @@
+const Pattern = require('./pattern.model');
```
Comment (Alan18081) [inline src/modules/pattern/pattern.service.js:1]: Please, move all newly added logic for patterns to another merge request to merge it in development as quickly as possible.

File: src/modules/user/user.service.js [hunk 20]
```diff
@@ -37,57 +37,47 @@ class UserService {
...
   }
 
-  getAllUsers() {
-    return User.find();
+  async getAllUsers() {
+    const users = await User.find();
+    if (users) {
+      return users;
+    }
+    throw new Error('Користувачів не знайдено');
```
Comment (Alan18081) [inline src/modules/user/user.service.js:45]: Even if no users are found, we should return empty array

File: src/utils/apolloClient.js [hunk 21]
```diff
@@ -0,0 +1,25 @@
+const { createHttpLink } = require('apollo-link-http');
```
Comment (Alan18081) [inline src/utils/apolloClient.js:1]: Please, rename this file to `apollo-client`

File: src/utils/validateUser.js [hunk 22]
```diff
@@ -28,14 +28,12 @@ exports.validateUpdateInput = Joi.object({
   firstName: Joi.string()
```
Comment (Alan18081) [inline src/utils/validateUser.js:28]: Please, move this change to the separate merge request, it is not related to the tests

**Comment by Alan18081** [comment]
Please, create separate merge requests:
- Tests for Users
